TypeScript: migrate server-side-render package to TS

kushagra-goyal-14 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Part of https://github.com/WordPress/gutenberg/issues/67691

<!-- In a few words, what is the PR actually doing? -->
Migrating the server-side-render package to TypeScript.
This PR supersedes #70755 as discussed in this PR

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
To enhance DX and add type safety.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
By porting the code and tests to TypeScript.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
Typecheck and unit tests.

kushagra-goyal-14 committed:
refactor: add server-side rendering functionality and related tests in ts

manzoorwanijk commented on packages/server-side-render/src/hook.ts:
```diff
@@ -145,7 +173,7 @@ export function useServerSideRender( args ) {
 							} );
 						} )
-						.catch( ( error ) => {
+						.catch( ( error: Error ) => {
```
The error is always unknown on runtime and that is how it should be as a type. If it's just have access to `error.name`, you can add `error instanceof Error` check before that. Otherwise, you can simply remove this type annotation here.

manzoorwanijk commented on packages/server-side-render/src/server-side-render.tsx:
```diff
@@ -1,3 +1,8 @@
+/**
+ * External dependencies
+ */
+import type { ComponentType, ReactNode } from 'react';
```
Can we please not import it here? We can directly use `React.ComponentType` or `React.ReactNode` without the need for any import

manzoorwanijk commented on packages/server-side-render/src/server-side-render.tsx:
```diff
@@ -16,27 +21,45 @@ import { useSelect } from '@wordpress/data';
 		// translators: %s: error message describing the problem
 		__( 'Error loading block: %s' ),
-		message
+		message || 'Unknown error'
```
This should be translated

```suggestion
		message || __( 'Unknown error' )
```

manzoorwanijk replied:
Another approach would be to get rid of this extra string by not making `message` optional in `ErrorPlaceholderProps` and then using `message={ error! }` in `<ErrorResponsePlaceholder />`. We can assert that because the description for `error` mentions `"The error message (available when status is 'error')."`, which is already the case when it's used there.

manzoorwanijk commented on packages/server-side-render/src/hook.ts:
```diff
@@ -5,17 +5,25 @@ import { debounce } from '@wordpress/compose';
 
-export function removeBlockSupportAttributes( attributes ) {
+export function removeBlockSupportAttributes(
+	attributes: Record< string, unknown >
```
Instead of making an assertion below for `attributes?.style`, we can type it here

```suggestion
	attributes: Record< string, unknown > & {
		style?: Record< string, unknown >;
	}
```

manzoorwanijk commented on packages/server-side-render/src/hook.ts:
```diff
@@ -35,7 +43,7 @@ export function removeBlockSupportAttributes( attributes ) {
 		typography,
 		...restStyles
-	} = attributes?.style || {};
+	} = ( attributes?.style as Record< string, unknown > ) || {};
```
We can then revert this change
```suggestion
	} = attributes?.style || {};
```

manzoorwanijk reviewed (commented):
This looks good. I have some suggestions for minor improvements

kushagra-goyal-14 committed:
Merge branch 'trunk' into try/migrate-server-side-render-to-ts

kushagra-goyal-14 committed:
Update README and TypeScript definitions for ServerSideRender component

- Enhanced README with detailed prop descriptions for ServerSideRender.
- Updated TypeScript types in hook.ts and server-side-render.tsx for better type safety and clarity.
- Improved error handling in useServerSideRender function.

manzoorwanijk commented on packages/server-side-render/src/server-side-render.tsx:
```diff
@@ -16,27 +10,45 @@ import { useSelect } from '@wordpress/data';
+
+interface LoadingPlaceholderProps {
+	children?: React.ReactNode;
+}
```
Let us extract these to an inline `types.ts` file.

manzoorwanijk commented on packages/server-side-render/src/server-side-render.tsx:
```diff
@@ -69,8 +81,15 @@ function DefaultLoadingResponsePlaceholder( { children } ) {
+	LoadingResponsePlaceholder?: React.ComponentType< LoadingPlaceholderProps >;
+}
+
+export function ServerSideRender( props: ServerSideRenderProps ) {
```
Let us also move this to `types.ts`.

manzoorwanijk commented on packages/server-side-render/src/server-side-render.tsx:
```diff
@@ -69,8 +81,15 @@ function DefaultLoadingResponsePlaceholder( { children } ) {
+}
+
+export function ServerSideRender( props: ServerSideRenderProps ) {
+	const [ prevContent, setPrevContent ] = useState( '' );
```
Why did we replace `useRef` with `useState`? 🤔

kushagra-goyal-14 replied:
Hi @manzoorwanijk `useRef` is correct conceptually here, but after converting to .tsx, the stricter eslint config enforces the `react-hooks/refs` rule which disallows reading ref.current during render and prevContentRef.current is read directly in JSX. That's why I was testing useState as an alternative.

I've reverted back to useRef and added` // eslint-disable-next-line react-hooks/refs` comments to suppress the lint error for now as an alternative, but let me know if you have any other suggestion for this. 🙌

manzoorwanijk commented on packages/server-side-render/src/server-side-render.tsx:
```diff
@@ -135,28 +157,24 @@ export function ServerSideRender( props ) {
+			| {
+					getCurrentPostId?: () => number | string | null;
+			  }
+			| undefined;
```
Instead of an assertion, it's probably better to have a `// @ts-expect-error` comment here.

manzoorwanijk commented on packages/server-side-render/src/server-side-render.tsx:
```diff
@@ -108,7 +125,12 @@ export function ServerSideRender( props ) {
+interface ServerSideRenderWithPostIdProps
+	extends Omit< ServerSideRenderProps, 'urlQueryArgs' > {
+	urlQueryArgs?: Record< string, unknown >;
 }
```
Let us move this to types.ts

manzoorwanijk reviewed (commented):
This looks pretty close. I have some minor suggestions and a question.

manzoorwanijk commented on packages/server-side-render/tsconfig.json:
```diff
@@ -0,0 +1,25 @@
+		"types": [ "gutenberg-env", "gutenberg-test-env", "jest" ]
+	},
+	"references": [
+		{ "path": "../api-fetch" },
```
```suggestion
		{ "path": "../api-fetch" },
		{ "path": "../blocks" },
```

manzoorwanijk commented on packages/server-side-render/tsconfig.json:
```diff
@@ -0,0 +1,25 @@
+		"src/**/react-native-*",
+		"src/**/stories/**/*.js",
+		"src/**/test/**/*.js"
+	]
```
Base config already does this, let us remove this whole `exclude` from here.

kushagra-goyal-14 committed:
refactor: Migrate ServerSideRender component to TypeScript and enhance type definitions

kushagra-goyal-14 committed:
Merge branch 'trunk' into try/migrate-server-side-render-to-ts

manzoorwanijk commented on packages/server-side-render/src/server-side-render.js:
````diff
@@ -135,28 +151,21 @@ export function ServerSideRender( props ) {
  * }
  * ```
  *
- * @param {Object}   props                                    Component props.
- * @param {string}   props.block                              The identifier of the block to be serverside rendered.
- * @param {Object}   props.attributes                         The block attributes to be sent to the server for rendering.
- * @param {string}   [props.className]                        Additional classes to apply to the wrapper element.
- * @param {string}   [props.httpMethod='GET']                 The HTTP method to use ('GET' or 'POST'). Default is 'GET'
- * @param {Object}   [props.urlQueryArgs]                     Additional query arguments to append to the request URL.
- * @param {boolean}  [props.skipBlockSupportAttributes=false] Whether to remove block support attributes before sending.
- * @param {Function} [props.EmptyResponsePlaceholder]         Component rendered when the API response is empty.
- * @param {Function} [props.ErrorResponsePlaceholder]         Component rendered when the API response is an error.
- * @param {Function} [props.LoadingResponsePlaceholder]       Component rendered while the API request is loading.
````
We have lost the JSDoc comments from these types. Can we please retain those descriptions in the types.ts files?

kushagra-goyal-14 replied:
Oh my bad, fixed it. Thanks

manzoorwanijk reviewed (changes requested):
Apart from the inline comment, we will need to add the `types` field in `package.json` and also in the `exports` field there.

kushagra-goyal-14 committed:
refactor: enhance type definitions for ServerSideRender component and include types in package.json

manzoorwanijk reviewed (approved):
This looks good now. Thank you
